### PR TITLE
[python] Replace bare `assert` with typed exceptions

### PR DESCRIPTION
**Pull request by johnkerl**
## Issue and/or context:

#433 

## Changes:

Switching to typed exceptions.

## Notes for Reviewer:

Along the way I realized the outgest needs programmable `X_layer_name` -- I'm happy to split this out into a separate PR if desired.

### CODE REVIEW

File: apis/python/src/tiledbsoma/collection.py [hunk 1]
```diff
@@ -183,7 +184,8 @@ def __delitem__(self, key: str) -> None:
     def __iter__(self) -> Iterator[str]:
         if self._cached_values is None:
             self._load_tdb_group_cache()
-        assert self._cached_values is not None
+        if self._cached_values is None:
```
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/collection.py:187]: `assert`s are fine for internal sanity checks, so in this situation (`assert self._cached_values is not None`) it is fine.

File: apis/python/src/tiledbsoma/collection.py [hunk 2]
```diff
@@ -356,7 +360,8 @@ def _tiledb_open(self, mode: str = "r") -> tiledb.Group:
         as ``with self._tiledb_open() as G:``
         as well as ``G = self._tiledb_open(); ...; G.close()``.
         """
-        assert mode in ("r", "w")
+        if mode not in ["r", "w"]:
+            raise SOMAError(f'expected mode to be one of "r" or "w"; got "{mode}"')
```
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/collection.py:364]: `ValueError`?

**Review by thetorpedodog — COMMENTED** [review]
This is mostly good; a few instances where it might make sense to use built-in exception types, and a few places where I would maintain `assert` statements.

File: apis/python/src/tiledbsoma/collection.py (hunk 1 above)
Comment (johnkerl) [reply apis/python/src/tiledbsoma/collection.py:187]: @thetorpedodog if this _ever_ goes off for a user I want a clear message "You didn't do anything wrong -- we did" surfaced to them. Hence the "internal error" wording.

**Comment by johnkerl** [comment]
@thetorpedodog ready for re-review, with comments for both your concerns

**Comment by johnkerl** [comment]
@mlin @atolopko-czi @thetorpedodog ping 🙏

File: apis/python/src/tiledbsoma/collection.py [hunk 3]
```diff
@@ -353,7 +357,8 @@ def _tiledb_open(self, mode: str = "r") -> tiledb.Group:
         as ``with self._tiledb_open() as G:``
         as well as ``G = self._tiledb_open(); ...; G.close()``.
         """
-        assert mode in ("r", "w")
+        if mode not in ["r", "w"]:
+            raise ValueError(f'expected mode to be one of "r" or "w"; got "{mode}"')
```
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/collection.py:361]: you might use `!r` to automatically `repr()` the value in question, so you need not bring your own quote marks:
```suggestion
            raise ValueError(f'expected mode to be one of "r" or "w"; got {mode!r}')
```

**Review by thetorpedodog — APPROVED** [review]
One other thing you can take or leave; otherwise looks good.

**Pull request merged**